xbmc/pvr/PVRActionListener.cpp

```diff
@@ -0,0 +1,158 @@
+              g_PVRManager.StartPlayback(PlaybackTypeTv);
+            break;
+          case ACTION_PVR_PLAY_RADIO:
+            if (!isPlayingPvr || !g_application.CurrentFileItem().GetPVRChannelInfoTag()->IsRadio())
+              g_PVRManager.StartPlayback(PlaybackTypeRadio);
+            break;
+        }
+      }
+      return true;
+    }
+    case REMOTE_0:
+    case REMOTE_1:
+    case REMOTE_2:
+    case REMOTE_3:
+    case REMOTE_4:
+    case REMOTE_5:
+    case REMOTE_6:
+    case REMOTE_7:
+    case REMOTE_8:
+    case REMOTE_9:
+    {
+      if (g_application.IsFullScreen() && g_application.CurrentFileItem().IsLiveTV())
+      {
+        if(g_PVRManager.IsPlaying())
+        {
+          // pvr client addon
+          CPVRChannelPtr playingChannel;
+          if(!g_PVRManager.GetCurrentChannel(playingChannel))
+            return false;
+          
+          if (action.GetID() == REMOTE_0)
+          {
+            CPVRChannelGroupPtr group = g_PVRChannelGroups->GetPreviousPlayedGroup();
+            if (group)
+            {
+              g_PVRManager.SetPlayingGroup(group);
+              CFileItemPtr fileItem = group->GetLastPlayedChannel(playingChannel->ChannelID());
+              if (fileItem && fileItem->HasPVRChannelInfoTag())
+              {
+                CLog::Log(LOGDEBUG, "%s - switch to channel number %d", __FUNCTION__, fileItem->GetPVRChannelInfoTag()->ChannelNumber());
+                g_application.OnAction(CAction(ACTION_CHANNEL_SWITCH, (float) fileItem->GetPVRChannelInfoTag()->ChannelNumber()));
+              }
+            }
+          }
+          else
+          {
+            int autoCloseTime = CSettings::Get().GetBool("pvrplayback.confirmchannelswitch") ? 0 : g_advancedSettings.m_iPVRNumericChannelSwitchTimeout;
+            std::string strChannel = StringUtils::Format("%i", action.GetID() - REMOTE_0);
+            if (CGUIDialogNumeric::ShowAndGetNumber(strChannel, g_localizeStrings.Get(19000), autoCloseTime) || autoCloseTime)
+            {
+              int iChannelNumber = atoi(strChannel.c_str());
+              if (iChannelNumber > 0 && iChannelNumber != playingChannel->ChannelNumber())
+              {
+                CPVRChannelGroupPtr selectedGroup = g_PVRManager.GetPlayingGroup(playingChannel->IsRadio());
+                CFileItemPtr channel = selectedGroup->GetByChannelNumber(iChannelNumber);
+                if (!channel || !channel->HasPVRChannelInfoTag())
+                  return false;
+                
+                g_application.OnAction(CAction(ACTION_CHANNEL_SWITCH, (float)iChannelNumber));
+              }
+            }
+          }
+        }
+        else
+        {
+          // filesystem provider like slingbox, cmyth, etc
+          int iChannelNumber = -1;
+          std::string strChannel = StringUtils::Format("%i", action.GetID() - REMOTE_0);
+          if (CGUIDialogNumeric::ShowAndGetNumber(strChannel, g_localizeStrings.Get(19000)))
```

Comment:
if the vector is protected other components cant register/unregister. and you have to brag the lock while calling OnAction to prevent the component from unregistering while in OnAction